apm-sniffer/apm-sdk-plugin/solr-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/solr/shard/ShardRequestInterceptor.java

```diff
@@ -0,0 +1,110 @@
+
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import static org.apache.skywalking.apm.plugin.solr.commons.Constants.*;
+
+public class ShardRequestInterceptor implements InstanceMethodsAroundInterceptor, InstanceConstructorInterceptor {
+
+    @Override
+    public void onConstruct(EnhancedInstance objInst, Object[] allArguments) {
+        HttpSolrClient client = (HttpSolrClient) objInst;
+        String peer = "", core = "";
+        try {
+            URL url = new URL(client.getBaseURL());
+            peer = url.getHost() + ":" + url.getPort();
+            int idx = url.getPath().lastIndexOf('/');
+            if (idx > 0) {
+                core = url.getPath().substring(idx);
+            }
+        } catch (MalformedURLException ignore) {
+        }
+        objInst.setSkyWalkingDynamicField(new String[]{peer, core});
+    }
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        SolrRequest<?> request = (SolrRequest<?>) allArguments[0];
+        if (request.getPath().startsWith(SKIP_ADMIN_PREF))
+            return;
+
+        String[] info = (String[]) objInst.getSkyWalkingDynamicField();
+        SolrParams params = request.getParams();
+
+        ContextSnapshot snapshot = null;
+        String operationName = OPER_SHARD_PREF + info[1] + request.getPath();
+        if (params.getBool(SW_ENHANCE_FLAG, false)) {
+            if (params instanceof ModSolrParamsWrapper) {
+                EnhancedInstance instance = (EnhancedInstance) request.getParams();
+                snapshot = (ContextSnapshot) instance.getSkyWalkingDynamicField();
+            }
+            ((ModifiableSolrParams) params).remove(SW_ENHANCE_FLAG);
+            final int purpose = params.getInt(ShardParams.SHARDS_PURPOSE, 0);
+            operationName = operationName + NamesMap.gePurposeName(purpose);
+        }
+
+        AbstractSpan span = ContextManager.createExitSpan(operationName, info[0]);
+        span.setComponent(ComponentsDefine.SOLR);
+        span.setLayer(SpanLayer.HTTP);
+
+        if (snapshot != null) {
+            ContextManager.continued(snapshot);
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                              Object ret) throws Throwable {
+        SolrRequest<?> request = (SolrRequest<?>) allArguments[0];
+        if (!request.getPath().startsWith(SKIP_ADMIN_PREF) && ContextManager.isActive()) {
+            ContextManager.stopSpan();
+        }
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t) {
+        if (ContextManager.isActive()) {
```

Comment:
`!request.getPath().startsWith(SKIP_ADMIN_PREF)` should be here

Reply:
Yes, I notice that. Maybe it is not required. In usually, there is no active span which matches this condition here. (Unless other interceptors create.)

That there is a active span in context, not matter of who created, logs the exception when it is in the method with catching exception. I think.